packages/ics721/src/ibc_packet_receive.rs

```diff
@@ -62,25 +31,83 @@ pub(crate) fn receive_ibc_packet(
     let data: NonFungibleTokenPacketData = from_json(&packet.data)?;
     data.validate()?;
 
-    let cloned_data = data.clone();
-    let receiver = deps.api.addr_validate(&data.receiver)?;
-    let token_count = data.token_ids.len();
-
     // Check if NFT is local if not get the local class id
     let maybe_local_class_id = try_pop_source_prefix(&packet.src, &data.class_id);
     let callback = get_receive_callback(&data);
+    let local_class_id = if let Some(local_class_id) = maybe_local_class_id {
+        ClassId::new(local_class_id)
+    } else {
+        let local_prefix = get_endpoint_prefix(&packet.dest);
+        ClassId::new(format!("{}{}", local_prefix, data.class_id))
+    };
 
-    let action_aggregator = data
+    // sub message holds 2 to 4 messages:
+    // - one message for voucher creation or redemption, another message for updating incoming or outgoing channel
+    let (is_redemption, voucher_and_channel_messages) = create_voucher_and_channel_messages(
+        deps.as_ref(),
+        env.clone(),
+        data.clone(),
+        maybe_local_class_id,
+        local_class_id.clone(),
+        packet.clone(),
+    )?;
+    // - one optional incoming proxy message
+    let incoming_proxy_msg =
+        get_incoming_proxy_msg(deps.as_ref().storage, packet.clone(), data.clone())?;
+    // - one optional callback message
+    let callback_msg = create_callback_msg(
+        deps.as_ref(),
+        &env,
+        &data,
+        is_redemption,
+        callback,
+        local_class_id,
+    )?;
+
+    let submessage = into_submessage(
+        env.contract.address,
+        voucher_and_channel_messages.0,
+        voucher_and_channel_messages.1,
+        callback_msg,
+        incoming_proxy_msg,
+    )?;
+
+    let response = if let Some(memo) = data.memo {
+        IbcReceiveResponse::default().add_attribute("ics721_memo", memo)
+    } else {
+        IbcReceiveResponse::default()
+    };
+
+    Ok(response
+        .add_submessage(submessage)
+        .add_attribute("method", "receive_ibc_packet")
+        .add_attribute("class_id", data.class_id)
+        .add_attribute("local_channel", packet.dest.channel_id)
+        .add_attribute("counterparty_channel", packet.src.channel_id))
+}
+
+fn create_voucher_and_channel_messages(
+    deps: Deps,
+    env: Env,
+    data: NonFungibleTokenPacketData,
+    maybe_local_class_id: Option<&str>,
+    local_class_id: ClassId,
+    packet: IbcPacket,
+) -> Result<(bool, (WasmMsg, WasmMsg)), ContractError> {
+    let token_count = data.token_ids.len();
+    let redemption_or_create = data
         .token_ids
         .into_iter()
         .zip_optional(data.token_uris)
         .zip_optional(data.token_data)
         .try_fold(
-            Vec::<Action>::with_capacity(token_count),
-            |mut messages, ((token_id, token_uri), token_data)| -> StdResult<_> {
```

Comment:
old logic is super complicated for create a  callaback submessage, using Action and Action Aggregator